pkg/provider/cache/cache.go

```diff
@@ -0,0 +1,99 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+// Cache is a generic lru cache that allows you to
+// lookup values using a key and a version.
+// By design, this cache allows access to only a single version of a given key.
+// A version mismatch is considered a cache miss and the key gets evicted if it exists.
+// When a key is evicted a optional cleanup function is called.
+type Cache[T any] struct {
+	lru         *lru.Cache
+	size        int
+	cleanupFunc cleanupFunc[T]
```

Comment:
We allow this mutator as a way to control what happens when the session expires, right?. If that's so, wouldn't it be worth to also foresee a Get Mutator as a way to renew the lease and not allowing it to expire? Not saying to implement it now, just to have it in mind.

Reply:
> when the session expires

Not when the session expire; It is called when the LRU reaches max capacity and evicts the last-recently-used item from the cache.

I see where you're heading with your questions, we can consider building something on top of this that provides session management capabilities.
But i'm not sure yet how that could look like (do we want to persist that somewhere? what happens when a controller restarts? etc.)